stan/math/prim/prob/wiener4_lccdf.hpp

```diff
@@ -0,0 +1,454 @@
+#ifndef STAN_MATH_PRIM_PROB_WIENER4_LCCDF_HPP
+#define STAN_MATH_PRIM_PROB_WIENER4_LCCDF_HPP
+
+#include <stan/math/prim/prob/wiener4_lcdf.hpp>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * Log of probability of reaching the upper bound in diffusion process
+ *
+ * @tparam T_a type of boundary
+ * @tparam T_w type of relative starting point
+ * @tparam T_v type of drift rate
+ *
+ * @param a The boundary separation
+ * @param w_value The relative starting point
+ * @param v_value The drift rate
+ * @return log probability to reach the upper bound
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto wiener_prob(const T_a& a, const T_v& v_value,
+                        const T_w& w_value) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  const auto v = -v_value;
+  const auto w = 1 - w_value;
+  if (fabs(v) == 0.0) {
+    return ret_t(log1p(-w));
+  }
+
+  const auto exponent = -2.0 * v * a * (1.0 - w);
+  if (exponent < 0) {
+    return ret_t(log1m_exp(exponent) - log_diff_exp(2 * v * a * w, exponent));
+  } else {
+    return ret_t(log1m_exp(-exponent) - log1m_exp(2 * v * a));
+  }
+}
+
+/**
+ * Calculate parts of the partial derivatives for wiener_prob_grad_a and
+ * wiener_prob_grad_v (on log-scale)
+ *
+ * @tparam T_a type of boundary
+ * @tparam T_w type of relative starting point
+ * @tparam T_v type of drift rate
+ *
+ * @param a The boundary separation
+ * @param w_value The relative starting point
+ * @param v_value The drift rate
+ * @return 'ans' term
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto wiener_prob_derivative_term(const T_a& a, const T_v& v_value,
+                                        const T_w& w_value) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  const auto exponent_m1 = log1p(-1.1 * 1.0e-8);
```

Comment:
Where does this hard coded value come from?

Reply:
Here, I have to correct myself. 
This term serves also numerical stability. We test whether the exponents are larger than this small value. If this is the case, then they are negative and very near to 0. In the limit, when the exponents go to 0, the result is `-w`. As we later have to divide by the exponents, we would have to divide by nealry zero. Therefore, we do this check and return `-w` for exponents very near to zero to make compuatations more stable.